base/applications/rapps/applicationdb.cpp

```diff
@@ -0,0 +1,292 @@
+    const CAtlList<CApplicationInfo *> &list = UseInstalled ? m_Installed : m_Available;
+    const BOOL IncludeAll = UseInstalled ? (Type == ENUM_ALL_INSTALLED) : (Type == ENUM_ALL_AVAILABLE);
+
+    POSITION CurrentListPosition = list.GetHeadPosition();
+    while (CurrentListPosition)
+    {
+        CApplicationInfo *Info = list.GetNext(CurrentListPosition);
+
+        if (IncludeAll || Type == Info->iCategory)
+        {
+            List.AddTail(Info);
+        }
+    }
+}
+
+
+BOOL CApplicationDB::EnumerateFiles()
+{
+    ClearList(m_Available);
+
+    CPathW AppsPath = m_BasePath;
+    AppsPath += L"rapps";
+    CPathW WildcardPath = AppsPath;
+    WildcardPath += L"*.txt";
+
+    WIN32_FIND_DATAW FindFileData;
+    HANDLE hFind = FindFirstFileW(WildcardPath, &FindFileData);
+    if (hFind == INVALID_HANDLE_VALUE)
+    {
+        return FALSE;
+    }
+
+    do
+    {
+        CStringW szPkgName = FindFileData.cFileName;
+        PathRemoveExtensionW(szPkgName.GetBuffer(MAX_PATH));
+        szPkgName.ReleaseBuffer();
+
+        CApplicationInfo* Info = FindByPackageName(szPkgName);
+        ATLASSERT(Info == NULL);
+        if (!Info)
+        {
+            CConfigParser* Parser = new CConfigParser(FindFileData.cFileName);
+            int Cat;
+            if (!Parser->GetInt(L"Category", Cat))
+                Cat = ENUM_INVALID;
+
+            Info = new CAvailableApplicationInfo(Parser, szPkgName, static_cast<AppsCategories>(Cat), AppsPath);
+            if (Info->Valid())
+            {
+                m_Available.AddTail(Info);
+            }
+            else
+            {
+                delete Info;
+            }
+        }
+
+    } while (FindNextFileW(hFind, &FindFileData));
+
+
+    FindClose(hFind);
+    return TRUE;
+}
+
+
+VOID CApplicationDB::UpdateAvailable()
+{
+    if (!CreateDirectoryW(m_BasePath, NULL) && GetLastError() != ERROR_ALREADY_EXISTS)
+        return;
+
+    if (EnumerateFiles())
+        return;
+
+    DownloadApplicationsDB(
+        SettingsInfo.bUseSource ? SettingsInfo.szSourceURL : APPLICATION_DATABASE_URL, !SettingsInfo.bUseSource);
+
+    CPathW AppsPath = m_BasePath;
+    AppsPath += L"rapps";
+    if (!ExtractFilesFromCab(APPLICATION_DATABASE_NAME, m_BasePath, AppsPath))
+        return;
+
+    CPathW CabFile = m_BasePath;
+    CabFile += APPLICATION_DATABASE_NAME;
+    DeleteFileW(CabFile);
+
+    EnumerateFiles();
+}
+
+VOID CApplicationDB::UpdateInstalled()
+{
+    // Remove all old entries
+    ClearList(m_Installed);
+
+    int LoopKeys = 2;
+
+    if (IsSystem64Bit())
+    {
+        // loop for all 3 combination.
+        // note that HKEY_CURRENT_USER\Software don't have a redirect
+        // https://docs.microsoft.com/en-us/windows/win32/winprog64/shared-registry-keys#redirected-shared-and-reflected-keys-under-wow64
+        LoopKeys = 3;
+    }
+
+    for (int keyIndex = 0; keyIndex < LoopKeys; keyIndex++)
+    {
+        DWORD dwSize = MAX_PATH;
```

Comment:
Maybe `DWORD dwSize = _countof(szKeyName);`  if you move the array declaration before that dwSize?